neumai/neumai/SinkConnectors/MarqoSink.py

```diff
@@ -0,0 +1,187 @@
+        if recreate_index:
+            if index_name in [i.index_name for i in marqo_client.get_indexes()['results']]:
+                marqo_client.delete_index(index_name)
+
+        if not index_name:
+            raise Exception('Index name must not be none')
+
+        if similarity not in ['cosinesimil', 'l1', 'l2', 'linf']:
+            raise Exception('Similarity function must be one of l1, l2, linf and cosinesimil')
+
+        if index_name in [i.index_name for i in marqo_client.get_indexes()['results']] and recreate_index==False:
+            raise Exception(f'Index {index_name} already exists, please use another name')
+
+        # Not using any model, providing own vectors
+        # https://docs.marqo.ai/1.4.0/API-Reference/Indexes/create_index/#no-model
+        marqo_client.create_index(
+        index_name=index_name,
+        settings_dict={
+                    'index_defaults': {
+                        'model': 'no_model',
+                        'model_properties': {
+                            'dimensions': embedding_dim
+                        },
+                        'ann_parameters':{
+                            'space_type': similarity
+                        }
+                    }
+                }
+        )
+
+    def store(self, vectors_to_store:List[NeumVector]) -> int:
+        url = self.url
+        api_key = self.api_key
+        index_name = self.index_name
+
+        marqo_client = marqo.Client(
+            url=url, 
+            api_key=api_key,
+        )
+        self._create_index(index_name=index_name,
+                          marqo_client=marqo_client,
+                          similarity="cosinesimil",
+                          embedding_dim=len(vectors_to_store[0].vector))
+        mod_vecs = []
+        for vec in vectors_to_store:
+            _id = vec.id
+            _vec = vec.vector
+            dic = {
+                    '_id': _id,
+                    'neum': {
+                        'vector': _vec
+                    }
+                }
+            # Using the two lines below, we can unroll the elements in metadata
+            # and add them separately
+            for k,v in vec.metadata.items():
+                dic[k] = v
+            # dic['metadata'] = vec.metadata
+            mod_vecs.append(
+                dic
+            )
+        operation_info = marqo_client.index(index_name).add_documents(
+            documents=mod_vecs,
+            mappings={
+                'neum': 
+                    {
+                    'type': 'custom_vector'
+                    }
+            },
+            tensor_fields=['neum'],
+            auto_refresh=True
+            )
+        
+        if(operation_info['errors'] == False):
+            return  len(operation_info['items'])
+        raise MarqoInsertionException("Marqo storing failed. Try again later.")
+    
+    def search(self, vector: List[float], number_of_results: int, filter:dict = {}) -> List:
```

Comment:
Looks like Marqo support query filtering but it is done with a string, not a dict. Would be great to support potentially.

Reply:
Also, can you show a sample filter dictionary, how it looks like?

Reply:
Currently the filter dictionary is based on the format specific for a given vector DB. We haven't implemented a unified language.

For example, Pinecone uses something similar to MongoDB, ex. {"field1":"value"}

Reply:
Okay @ddematheu , but I guess in this setting, we cannot have range based queries. In future its better to have a way for specifying and mapping filters from different destinations, for example as given [here](https://github.com/mindsdb/mindsdb/blob/e594c79c294ef8e64ff250af01e6f35148e3547b/mindsdb/integrations/libs/vectordatabase_handler.py#L52), we can define a structure for filterconditions with a `field`, `operation` and `value`. 

So, for now, I will implement based on what you said - `{"field1": "value"}`.

Reply:
I agree on the need for a unified filter condition mapping. Will add it as an issue to work on next.